test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go

```diff
@@ -0,0 +1,459 @@
+	clusterToDockerMachinePools, err := util.ClusterToObjectsMapper(mgr.GetClient(), &infrav1exp.DockerMachinePoolList{}, mgr.GetScheme())
+	if err != nil {
+		return err
+	}
+
+	c, err := ctrl.NewControllerManagedBy(mgr).
+		For(&infrav1exp.DockerMachinePool{}).
+		WithOptions(options).
+		WithEventFilter(predicates.ResourceNotPaused(r.Log)).
+		Watches(
+			&source.Kind{Type: &clusterv1exp.MachinePool{}},
+			&handler.EnqueueRequestsFromMapFunc{
+				ToRequests: utilexp.MachinePoolToInfrastructureMapFunc(infrav1.GroupVersion.WithKind("DockerMachinePool")),
+			},
+		).
+		Watches(
+			&source.Kind{Type: &infrav1.DockerCluster{}},
+			&handler.EnqueueRequestsFromMapFunc{
+				ToRequests: handler.ToRequestsFunc(r.DockerClusterToDockerMachinePools),
+			},
+		).
+		Build(r)
+	if err != nil {
+		return err
+	}
+	return c.Watch(
+		&source.Kind{Type: &clusterv1.Cluster{}},
+		&handler.EnqueueRequestsFromMapFunc{
+			ToRequests: clusterToDockerMachinePools,
+		},
+		predicates.ClusterUnpausedAndInfrastructureReady(r.Log),
+	)
+}
+
+// DockerClusterToDockerMachines is a handler.ToRequestsFunc to be used to enqueue
+// requests for reconciliation of DockerMachines.
+func (r *DockerMachinePoolReconciler) DockerClusterToDockerMachinePools(o handler.MapObject) []ctrl.Request {
+	result := []ctrl.Request{}
+	c, ok := o.Object.(*infrav1.DockerCluster)
+	if !ok {
+		r.Log.Error(errors.Errorf("expected a DockerCluster but got a %T", o.Object), "failed to get DockerMachinePool for DockerCluster")
+		return nil
+	}
+	log := r.Log.WithValues("DockerCluster", c.Name, "Namespace", c.Namespace)
+
+	cluster, err := util.GetOwnerCluster(context.TODO(), r.Client, c.ObjectMeta)
+	switch {
+	case apierrors.IsNotFound(err) || cluster == nil:
+		return result
+	case err != nil:
+		log.Error(err, "failed to get owning cluster")
+		return result
+	}
+
+	labels := map[string]string{clusterv1.ClusterLabelName: cluster.Name}
+	machinePoolList := &clusterv1exp.MachinePoolList{}
+	if err := r.Client.List(context.TODO(), machinePoolList, client.InNamespace(c.Namespace), client.MatchingLabels(labels)); err != nil {
+		log.Error(err, "failed to list DockerMachinePools")
+		return nil
+	}
+
+	for _, m := range machinePoolList.Items {
+		if m.Spec.Template.Spec.InfrastructureRef.Name == "" {
+			continue
+		}
+		name := client.ObjectKey{Namespace: m.Namespace, Name: m.Name}
+		result = append(result, ctrl.Request{NamespacedName: name})
+	}
+
+	return result
+}
+
+func (r *DockerMachinePoolReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, machinePool *clusterv1exp.MachinePool, dockerMachinePool *infrav1exp.DockerMachinePool, log logr.Logger) (ctrl.Result, error) {
+	// check if there are more machines than the spec requires
+	realMachineCount := int32(len(dockerMachinePool.Status.Machines))
+	if realMachineCount > 0 && realMachineCount > *machinePool.Spec.Replicas {
```

Comment:
In the case of delete, I was expecting the controller to scale down to 0, while instead, it is scaling down to `Spec.Replicas`. Am I missing something?

Reply:
Delete now removes all docker machines labeled with the machine pool.